crates/bootstrap_srv/src/space.rs

```diff
@@ -0,0 +1,228 @@
+                first = false;
+            } else {
+                out.push(b',');
+            }
+            r.read(&mut out)?;
+        }
+
+        out.push(b']');
+
+        debug_assert_eq!(len, out.len());
+
+        Ok(out)
+    }
+
+    /// Update a space, clearing out any expired infos and optionally
+    /// adding a new incoming info. If adding a new info, perform all
+    /// validation before calling this function.
+    pub fn update(
+        &self,
+        max_entries: usize,
+        mut new_info: Option<(crate::ParsedEntry, Option<StoreEntryRef>)>,
+    ) {
+        // get the current system time
+        let now = std::time::SystemTime::now()
+            .duration_since(std::time::SystemTime::UNIX_EPOCH)
+            .expect("InvalidSystemTime")
+            .as_micros() as i64;
+
+        // only allow one write at a time
+        let write_lock = self.write_lock.lock().unwrap();
+
+        // keep the readable lock time to a minimum
+        let list = self.readable.lock().unwrap().clone();
+
+        // if we have a new info to add, pull out the agent key
+        // so we can check to replace an existing one while
+        // iterating the existing infos
+        let replace_agent: Option<ed25519_dalek::VerifyingKey> =
+            new_info.as_ref().map(|(p, _)| p.agent);
+
+        // parse the current entries
+        let mut list = list
+            .into_iter()
+            .filter_map(|mut store| {
+                let mut parsed = match store.parse() {
+                    Ok(p) => p,
+                    Err(_) => return None,
+                };
+
+                // if this matches an existing info, and the new one is newer
+                // use the new one. note, we'll still check for expiration
+                // after this step.
+                if Some(parsed.agent) == replace_agent {
+                    // can unwrap because this can only be true if is_some
+                    let (new_p, new_s) = new_info.take().unwrap();
+                    if new_p.created_at > parsed.created_at {
+                        if new_p.is_tombstone {
+                            return None;
+                        }
+
+                        parsed = new_p;
+                        store = new_s.expect("RequireStoreRef");
+                    }
+                }
+
+                // check for expiration
+                if parsed.expires_at <= now {
+                    return None;
+                }
+
+                Some((parsed, store))
+            })
+            .collect::<Vec<_>>();
+
+        // if we still have a new info to add, it must not have matched
+        // any existing ones
+        if let Some((parsed, store)) = new_info {
+            if !parsed.is_tombstone {
+                // if we are full, follow the "default" strategy rule of
+                // deleting the half-way info
+                if list.len() >= max_entries {
+                    list.remove(max_entries / 2);
```

Comment:
would it be clearer to have an explicit "strategy" defined somewhere and this function to call it?

Reply:
For sure eventually this will need to change. But I'm not sure the scope of the change that will be required until we go to actually design the new strategy... so I'm inclined to keep this simple until needed.